### ISSUE TITLE
ASA All Contexts via Admin Context Login

### ISSUE THREAD

**Issue by jthunt** [ytti/oxidized#338]
Anyone have a good working model to save the admin, system, and all other contexts on a multi context ASA?

Thanks!

**Comment by etern4ever** [comment]
Well..  I was bored and hacked something together.  Not knowing Ruby that well, I really doubt I did this correctly.  It's a bad hack but it works for me.  Someone better at Ruby coding please fix this and add it to the models available in the repo?

Created a new model using the existing ASA and changed the Model Name.
Removed the show running section and replaced it with the following.

` cmd 'changeto system' do |cfg|
   cmd 'show running-config' do |systemcfg|
      allcfg = "\n\n" + systemcfg + "\n\n"
      contexts = systemcfg.scan(/^context (\S+)$/)
      contexts.each do |context|
         cmd "changeto context " + context.join(" ") do
           cmd 'show running-config' do |cfgcontext|
              allcfg = allcfg + "\n\n" + cfgcontext
           end
         end
      end
      cfg = allcfg
   end
   cfg
 end`

**Comment by yostie** [comment]
I didn't notice this issue and posted a request for multi-context ASA support.  Not sure how bad your code is because I don't know Ruby, but the only thing I would suggest is don't do a "show running-config", instead do a "more disk0:/context_name".  This way you get some of the normally starred out sections of the config.  I may take your code and see how well it works for me.

**Comment by ytti** [comment]
I am happo accept pull request. I don't personally have ASAs, and don't have knowledge on them. So it might be prudent to try to ping the authors (i.e. @ytti) in ticket, to see if they want to chime in.

**Comment by yostie** [comment]
The code above by @etern4ever does work.  The only issue is that it dumps all contexts into a single file which is not desirable.  When it goes to obtain the current context config, it would be nice to write it to disk and then move on to the next one to do the same.  Making the changes to go to the system context and do a "more disk0:/context_name" would be trivial.  Unfortunately I don't know Ruby to have it write to disk although I can't imagine that would be difficult to add even with no knowledge of Ruby.  

I may try to find in what .rb file has the code for writing to disk and see what the config object looks like at that point.  If it was presented as a list of config objects, that would be easy to iterate over and write each one to disk.

**Comment by ytti** [comment]
Is #388 duplicate of this?

**Comment by panzerfly** [comment]
From my point of view, yes, #388 is a duplicate.

I've encountered the same issue having to backup ASA devices with both single and multiple context mode. My approach was not to create a new model (as basically it's the same ASA model) but to keep the same model and customize it so that it automatically detects what mode the ASA runs.

I've raised pull request #479 to merge the changes into the main code for this. I've already tested this against some devices (one with 41 contexts) and it works great. 

If anyone else could test so that we have more confirmations I'd be grateful.
